ui/app/components/app-form/project-settings.ts

```diff
@@ -0,0 +1,225 @@
+    return this.authCase === 4;
+  }
+
+  get git() {
+    return this.dataSource?.git;
+  }
+
+
+  validateGitUrl() {
+    let gitUrl = new URL(this.project.dataSource.git.url)
+    // If basic auth, match https url
+    if (this.authCase == 4) {
+      if (gitUrl.protocol != 'https:') {
+        this.flashMessages.error('Git url needs to use "https:" protocol');
+        return false;
+      }
+    }
+    // If ssh force users to use a git: url
+    if (this.authCase == 5) {
+      if (gitUrl.protocol != 'git:') {
+        this.flashMessages.error('Git url needs to use "git:" protocol');
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @action
+  setAuthCase(val:any) {
+    this.authCase = val;
+  }
+
+  @action
+  setBasicAuth(path: string, e: any) {
+    if (!this.project.dataSource?.git?.basic) {
+      this.project.dataSource.git.basic = {
+        username: '',
+        password: ''
+      };
+    }
+    this.project.dataSource.git.basic[path] = e.target.value;
+  }
+
+  @action
+  setSshAuth(path: string, e: any) {
+    if (!this.project.dataSource?.git?.ssh) {
+      this.project.dataSource.git.ssh = {
+        user: '',
+        password: '',
+        privateKeyPem: '',
+      };
+    }
+    this.project.dataSource.git.ssh[path] = e.target.value;
+  }
+
+  @action
+  async saveSettings() {
+    if (!this.validateGitUrl()) {
+      return;
+    }
+    let project = this.project;
+    project.dataSource = this.dataSource;
+    let ref = new Project();
+    ref.setName(project.name);
+    let dataSource = new Job.DataSource();
+    let dataSourcePoll = new Project.Poll();
+    dataSourcePoll.setEnabled(project.dataSourcePoll.enabled);
+    dataSourcePoll.setInterval(project.dataSourcePoll.interval);
+    // Git settings
+    let git = new Job.Git();
+    git.setUrl(project.dataSource.git.url);
+    git.setPath(project.dataSource.git.path);
+    git.setRef(project.dataSource.git.ref);
+
+    // Git Authentication settings
+    if (this.authBasic) {
+      let gitBasic = new Job.Git.Basic();
+      gitBasic.setUsername(this.git.basic.username);
+      gitBasic.setPassword(this.git.basic.password);
+      git.setBasic(gitBasic);
+      git.clearSsh();
+    }
+
+    if (this.authSSH) {
+      let gitSSH = new Job.Git.SSH();
+      let encoder = new window.TextEncoder();
+      gitSSH.setPrivateKeyPem(encoder.encode(this.git.ssh.privateKeyPem));
+      gitSSH.setUser(this.git.ssh.user);
+      gitSSH.setPassword(this.git.ssh.password);
+      git.setSsh(gitSSH);
+      git.clearBasic();
+    }
+
+    dataSource.setGit(git);
+    ref.setDataSource(dataSource);
+    ref.setDataSourcePoll(dataSourcePoll);
+    let applist = project.applicationsList.map((app: Application.AsObject) => {
+      return new Application(app);
+    });
+    ref.setApplicationsList(applist);
+    let req = new UpsertProjectRequest();
+    req.setProject(ref);
+    await this.api.client
+      .upsertProject(req, this.api.WithMeta())
+      .then(() => {
```

Comment:
I think my favorite thing about `async/await` is you can use `try / catch` with this instead of `then().catch()`. It'd look like this:

```
try {
  await this.api.client.upsertProject(req, this.api.WithMeta());
  this.flashMessages.success('Settings saved');
  this.router.transitionTo('workspace.projects.project', this.project);
} catch (err) {
  this.flashMessages.error('Failed to save Settings', { content: err.message, sticky: true });
}
```